[gatsby-source-wordpress] support for exlcuded manufacturers and types

lightstrike opened the pull request:
I ran into some performance issues on a WP site with a ton of comments. Adding these parameters fixed the issues and are generally useful for tweaking the fetch behavior. I also reworked the valid routes function to be more explicit in its determinations of invalid routes. I didn't see any relevant tests - do other source plugins have tests written for fetch behavior?

Oh and added a line in gitignore to prevent vim swap files from sneaking in :smile_cat:

pieh commented on examples/using-wordpress/gatsby-config.js:
```diff
@@ -23,10 +23,43 @@ module.exports = {
+        // excludedManufacturers: [`disqus`, `yoast`],
+        // Exclude sepcific endpoint types
+        // Example: [`comments`] will exclude all `comments` type endpoints such as `/wp/v2/comments` from fetch. 
+        // excludedTypes: [`comments`, `media`],
```
Please don't change this config with example data - `README.md` is place for it - this is config for demo site - https://using-wordpress.gatsbyjs.org/ (I know that it currently doesn't build)

lightstrike replied:
Got it I'll remove.

pieh commented on packages/gatsby-source-wordpress/README.md:
```diff
@@ -83,11 +83,19 @@ plugins: [
+      // excludedManufacturers: [`disqus`, `yoast`],
+      // Exclude sepcific endpoint types
+      // Example: [`comments`] will exclude all `comments` type endpoints such as `/wp/v2/comments` from fetch.
+      // excludedTypes: [`comments`, `media`],
```
I feel like splitting exclusion into `excludedManufacturers` and `excludedTypes` limit flexibility of this feature. I can't really exclude `/acf/v3/posts` with this approach (either I exclude `acf` manufacturer and lose access to acf pages route or I exclude `posts` type and lose `/wp/v2/posts`). 

I would merge those into single `excludedRoutes` or `excludedEndpoints` and allow using globs (with [minimatch](https://www.npmjs.com/package/minimatch) package) or regex there so I could define them like this:
`excludedRoutes: ['acf/posts', '*/comments', 'yoast/*']`

Thoughts?

Also uncomment lines where you show how to exclude things. We want actual options to be quickly visible and now they blend with comments.

lightstrike replied:
I agree this is a more flexible approach, I'll update shortly.

lightstrike commented:
By the way @pieh, is there an easy way to install a fork of a gatsby plugin? 

I tried this: `npm install --save https://github.com/lightstrike/gatsby/tarball/wordpress/excluded-manufacturers-types` 

But hit this error: `Can't install git+https://github.com/lightstrike/gatsby.git#6705041cdca702cc2037cf9a50cdbbc817c59780: Missing package name`

I think it has to do with how the gatsby-source-wordpress package is within a subdirectory. Any thoughts? I'd like to test and potentially use on a live project while this PR is worked through (and ideally do so before opening PRs in the future).

pieh commented:
@lightstrike I pasted this link on discord - but in case You've missed it - check https://www.gatsbyjs.org/docs/how-to-contribute/#contributing-to-the-repo and go through bullet points

lightstrike commented:
@pieh Took a first shot at the single `excludedRoutes` option. Haven't done a lot of globbing and feel the current solution is quite fragile. Couldn't get it to work with ending endpoints such as `*/comments` as well as entire resources like `yoast/*`.

Any suggestions how to improve support for your example: `excludedRoutes: ['acf/posts', '*/comments', 'yoast/*']`?

pieh commented:
This was just example/concept what I think should be possible with this feature - it doesn't need to be in exact way I described.

I think reason it doesn't work in your current implementation is because You use link (`route._links.self` which contains `/vX/` version in the path as well as host and `wp-json` part). So this should be normalized to either `manufacturer/version/endpoint` or `manufacturer/endpoint` before trying to match with globs (currently I think variant with version would be better).

I tested with http://pthrasher.github.io/minimatch-test/ against `manufacturer/version/endpoint` variant and used globs seem to work well if we will get normalized paths to test against
![minimatch_test](https://user-images.githubusercontent.com/419821/37465724-3910c924-285c-11e8-8d3d-7c27f8a659ba.jpg)

lightstrike commented:
@pieh pushed an update that feels a lot cleaner. The big issue I wasn't able to figure out is how get all the patterns to match within an extglob. Any ideas here?

![screenshot from 2018-03-15 09-09-37](https://user-images.githubusercontent.com/1368742/37475908-081dbdb6-2831-11e8-8348-2695c22da0d4.png)

![screenshot from 2018-03-15 09-12-07](https://user-images.githubusercontent.com/1368742/37475924-0f0ec958-2831-11e8-8ed6-36a08f471896.png)

pieh commented:
You don't have to construct single pattern from all excludedRoutes - you can just iterate _excludedRoutes array and see if any of the routes matches using https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some
```diff
-} else if (minimatch(routePath, excludedRoutes)) {
+} else if (_excludedRoutes.some(excludedRoute => minimatch(routePath, excludedRoute))) {
```

lightstrike commented:
Great suggestion :smile: I think all is set to go now, let me know if there's anything else you'd like me to do. Has there been any kind of test harness written for the fetch logic in other source plugins?

pieh commented:
I don't think any source plugin test actual fetching part. But I think it would be possible to do by mocking `axios` to return predefined test data.

If you would like to continue working on improving performance of this plugin You could take look at is parallelizing fetching - right now it's done sequentially but it could use for example https://www.npmjs.com/package/better-queue to run few requests in parallel.

lightstrike commented:
@pieh `package.json` updated :+1: 

[axios-mock-adapter](https://www.npmjs.com/package/axios-mock-adapter) looks promising, I'll investigate that.

[better-queue](https://www.npmjs.com/package/better-queue) looks great, I'll work on that. I'm working on getting some big clunky WP sites onto Gatsby, where it would help to actually slow down fetching. I'll see if I can get some options together for people to be able to tweak faster/slower depending on their upstream conditions. :smile: 

By the way, re: installing a package fork inside a monorepo, I ended up creating [a standalone repo](https://github.com/lightstrike/gatsby-source-wordpress) and then running `npm install https://github.com/lightstrike/gatsby-source-wordpress/tarball/master`, which saved in `package.json` as `"gatsby-source-wordpress": "https://github.com/lightstrike/gatsby-source-wordpress/tarball/master",` -- it definitely works, though it'd still be nice to know how to install from inside the monorepo fork.

pieh commented:
For slowing things down you can check rate limiting / throttling (for example implemented here - https://github.com/LeKoArts/gatsby-source-behance/blob/master/src/gatsby-node.js#L53-L75 )

lightstrike commented:
@pieh anything else you'd like me to do on this?

I plan on addressing tests and adding a queue in other PRs.

pieh commented:
From my side, this is fine. I was just waiting on more input from other people.

@sebastienfi did You have time to look at this?

KyleAMathews commented:
LGTM

lightstrike commented:
Hey all, I just rebased my PR based on the excellent work by @tsimons on #4735 - it's helped me already :smile: 

The last commit has some cleanup on #4735 I'd recommend:
* renaming `concurrentRequests` to `_concurrentRequests` to maintain consistency of prefixing parameters passed by `gatsby-config.js` with an underscore
* documenting `concurrentRequests` in the readme
* fixed a bug on the [getPages call](https://github.com/gatsbyjs/gatsby/pull/4538/commits/ca31266b23478a527c2c93b39a69f84a49b3c31d#diff-16b936059133e8f2fe251106655a7d1bR210) (was wondering why I wasn't seeing this data before)

lightstrike commented:
Oh another thing I just thought of is to add better-queue to package.json


On Tue, Mar 27, 2018, 11:06 AM Kyle Mathews <notifications@github.com>
wrote:

> LGTM
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gatsbyjs/gatsby/pull/4538#issuecomment-376601355>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABTipv-fujD0R1kXYPtTKKOZHajMJ0sSks5tinGfgaJpZM4SrQBq>
> .
>

KyleAMathews commented:
Preview built from this PR https://deploy-preview-4538--using-wordpress.netlify.com/

lightstrike commented:
Thanks @KyleAMathews, two questions for you about deploy previews:

1. Do you know what causes [these kinds of deploy preview fails](https://app.netlify.com/sites/using-wordpress/deploys/5abbc42ffd0efa77cc00680b)?
2. What are your thoughts on #4608?

sebastienfi commented:
> Do you know what causes these kinds of deploy preview fails?

Sometimes NPM servers are unable to fullfil a request. Investigated with Netlify Team, seems to happen to a almost everyone using NPM and is independant from their infrastructure. Once the NPM package is in their cache this doesn't happen anymore.
Automatically retrying a failed build isn't a way to go either, so I guess building in serverless still requires a bit of human interaction.

jlengstorf commented:
Hiya @lightstrike! 👋

This is definitely late, but on behalf of the entire Gatsby community, I wanted to say thank you for being here.

Gatsby is built by awesome people like you. Let us say “thanks” in two ways:

1.  **We’d like to send you some Gatsby swag.** As a token of our appreciation, you can go to the [Gatsby Swag Store][store] and log in with your GitHub account to get a coupon code good for one free piece of swag. (We’ve got t-shirts and hats, plus some socks that are really razzing our berries right now.)
2.  **If you’re not already part of it, we just invited you to join the Gatsby organization on GitHub.** This will add you to our team of maintainers. You’ll receive an email shortly asking you to confirm. By joining the team, you’ll be able to label issues, review pull requests, and merge approved pull requests.

If you have questions, please don’t hesitate to reach out to us: tweet at [@gatsbyjs][twitter] and we’ll come a-runnin’.

Thanks again! 💪💜

[store]: https://store.gatsbyjs.org
[twitter]: https://twitter.com/gatsbyjs
